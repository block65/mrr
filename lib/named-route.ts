import type { Params, ExtractRouteParams } from './types.js';
import { nullOrigin, urlObjectAssign, urlRhs } from './util.js';

type QueryParams = Record<string, string>;

export interface NamedRoute<P extends string, Q extends QueryParams> {
  path: P;
  build: (x?: {
    params?: ExtractRouteParams<P>;
    query?: Q;
    hash?: string;
    origin?: string;
  }) => string;
}

export function interpolate<P extends string>(
  path: P,
  params: ExtractRouteParams<P> | Params = {},
): string {
  return path.replace(/\/:(\w+)[?+*]?/g, (_match, token: keyof typeof params) =>
    params[token] ? `/${params[token]}` : '',
  );
}

export function namedRoute<T extends string, Q extends QueryParams = never>(
  pathname: T,
): NamedRoute<T, Q> {
  return {
    path: pathname,
    build({ params, query, origin, ...rest } = {}) {
      const interpolatedPathname = params
        ? interpolate(pathname, params)
        : pathname;

      const searchParams = new URLSearchParams(query);
      searchParams.sort();

      const newUrl = urlObjectAssign(new URL(origin || nullOrigin), {
        pathname: interpolatedPathname,
        search: searchParams.toString(),
        ...rest,
      });

      return newUrl.origin === nullOrigin.origin
        ? urlRhs(newUrl)
        : newUrl.toString();
    },
  };
}
