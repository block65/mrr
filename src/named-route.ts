export { namedRoute } from '../lib/named-route.js';

export type {
  NamedRoute,
  NamedRouteWithParams,
  NamedRouteWithoutParams,
} from '../lib/named-route.js';
