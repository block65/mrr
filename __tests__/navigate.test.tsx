import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import type { FC } from 'react';
import { expect, test } from 'vitest';
import { Route, Router, Routes, useNavigate } from '../src/index.js';
import { namedRoute } from '../src/named-route.js';
import { NavigationInsideEffect } from './NavigationInsideEffect.js';
import { LocationDisplay } from './index.test.js';

const usersView = namedRoute('/users/:userId');

const Buttons: FC = () => {
  const { navigate } = useNavigate();

  const users = ['alice', 'bob', 'carol', 'dave'];

  return (
    <ul>
      {users.map((userId) => (
        <li key={userId}>
          <button
            type="button"
            data-testid={`button-${userId}`}
            onClick={() => {
              const href = usersView.build({ params: { userId } });
              navigate(href);
            }}
          >
            {userId}
          </button>
        </li>
      ))}
    </ul>
  );
};

test('nav with clicks', async () => {
  render(
    <Router pathname="/users/test1">
      <LocationDisplay />
      <Routes>
        <Route
          path={usersView.path}
          component={(params) => (
            <h1 data-testid={`heading-${params.userId}`}>
              You are user {params.userId}
            </h1>
          )}
        />

        <Route>
          <h1>fail</h1>
        </Route>
      </Routes>

      <Buttons />
    </Router>,
  );

  await waitFor(() => screen.getByTestId('heading-test1'));

  fireEvent.click(screen.getByTestId('button-alice'));
  await waitFor(() =>
    expect(screen.getByRole('heading')).toHaveTextContent(/alice/),
  );

  fireEvent.click(screen.getByTestId('button-bob'));
  await waitFor(() =>
    expect(screen.getByRole('heading')).toHaveTextContent(/bob/),
  );

  fireEvent.click(screen.getByTestId('button-carol'));
  await waitFor(() =>
    expect(screen.getByRole('heading')).toHaveTextContent(/carol/),
  );
});

test('programmatic navigation with hooks', async () => {
  render(
    <Router pathname="/">
      <NavigationInsideEffect />
    </Router>,
  );

  await waitFor(() => {
    expect(window?.location.href).toBe('http://localhost:3000/pickles?foo=bar');

    expect(screen.getByTestId('location-display')).toHaveTextContent(
      'http://localhost:3000/pickles?foo=bar',
    );
  });
});
