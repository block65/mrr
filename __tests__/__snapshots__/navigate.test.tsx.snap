// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
<DocumentFragment>
  <div
    data-testid="location-display"
  >
    http://localhost/login
  </div>
  <h1
    data-testid="login"
  >
    You are at the login page
  </h1>
</DocumentFragment>
`;
